### PR TITLE
am-eu-support

### PR DESCRIPTION
**Pull request by rohithkaza**
## Description of the change

> Providing the functionality in the way that existing customers can use the existing endpoint in the transformer.
> And provide an option in the dashboard which can be used for choosing between the endpoints.
> Option to opt for "EU Residency Server".
> By default the option should be set to the existing endpoint.

## Type of change
- [x] New feature (non-breaking change that adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Comment by akashrpo** [comment]
@rohithkaza Please describe the PR and fill in all its details.

File: v0/destinations/am/transform.js [hunk 1]
```diff
@@ -48,6 +48,11 @@ const logger = require("../../../logger");
 const AMBatchSizeLimit = 20 * 1024 * 1024; // 20 MB
 const AMBatchEventLimit = 500; // event size limit from sdk is 32KB => 15MB
 
+const EU_ENDPOINT = "https://api.eu.amplitude.com/2/httpapi";
+const EU_BATCH_EVENT_ENDPOINT = "https://api.eu.amplitude.com/batch";
+const EU_ALIAS_ENDPOINT = "https://api.eu.amplitude.com/usermap";
+const EU_GROUP_ENDPOINT = "https://api.eu.amplitude.com/groupidentify";
```
Comment (akashrpo) [inline v0/destinations/am/transform.js:54]: We should drive this from config.js

**Review by akashrpo — CHANGES_REQUESTED** [review]

**Comment by akashrpo** [comment]
The test case needs to be fixed `dataCenterEU` is missing from the destination.Config

**Review by ItsSudip — DISMISSED** [review]

File: v0/destinations/am/config.js [hunk 2]
```diff
@@ -113,6 +113,11 @@ const BATCH_EVENT_ENDPOINT = "https://api.amplitude.com/batch";
 const GROUP_ENDPOINT = "https://api.amplitude.com/groupidentify";
 const ALIAS_ENDPOINT = "https://api.amplitude.com/usermap";
 
+const EU_ENDPOINT = "https://api.eu.amplitude.com/2/httpapi";
+const EU_BATCH_EVENT_ENDPOINT = "https://api.eu.amplitude.com/batch";
+const EU_ALIAS_ENDPOINT = "https://api.eu.amplitude.com/usermap";
+const EU_GROUP_ENDPOINT = "https://api.eu.amplitude.com/groupidentify";
```
Comment (akashrpo) [inline v0/destinations/am/config.js:119]: EU endpoint for [identify](https://developers.amplitude.com/docs/identify-api#endpoint) is missing.

Comment (saikumarrs) [inline v0/destinations/am/config.js:119]: It's a bad idea to replicate all the constants for the EU server. It won't be extensible friendly. Instead, refactor these into a method.
I wouldn't insist on fixing it for the temporary release to the customer but should be addressed once eventually.
CC @utsabc

Comment (shrouti1507) [reply v0/destinations/am/config.js:119]: agreed. It will be more efficient.

Comment (shrouti1507) [reply v0/destinations/am/config.js:119]: agreed. It will be more efficient.

Comment (saikumarrs) [reply v0/destinations/am/config.js]: With the changes introduced, we have increased our chances of making mistakes down the line with the API endpoints. 
Here is a better way to implement it,

```
const BASE_URL = "https://api.amplitude.com";
const BASE_URL_EU = "https://api.eu.amplitude.com";

const baseEndPoint = (destConfig) => {
  var retVal;
  switch (destConfig.residencyServer) {
    case "EU":
      retVal = BASE_URL_EU;
      break;
    default: // "US" or when it is not specified
      retVal = BASE_URL;
  }
  return retVal;
}

const defaultEndPoint = (destConfig) => {
  var retVal = `${baseEndPoint(destConfig)}/2/httpapi`;
  return retVal;
}

const identifyEndPoint = (destConfig) => {
  var retVal = `${baseEndPoint(destConfig)}/identify`;
  return retVal;
}

const batchEndPoint = (destConfig) => {
  var retVal = `${baseEndPoint(destConfig)}/batch`;
  return retVal;
}

const groupEndPoint = (destConfig) => {
  var retVal = `${baseEndPoint(destConfig)}/groupidentify`;
  return retVal;
}

const aliasEndPoint = (destConfig) => {
  var retVal = `${baseEndPoint(destConfig)}/usermap`;
  return retVal;
}
```

Note that I've also changed the config variable representation to the choice of the residency server.

Comment (rohithkaza) [reply v0/destinations/am/config.js]: Nice implementation.
Updated as suggested.

File: v0/destinations/am/transform.js [hunk 3]
```diff
@@ -48,6 +46,44 @@ const logger = require("../../../logger");
 const AMBatchSizeLimit = 20 * 1024 * 1024; // 20 MB
 const AMBatchEventLimit = 500; // event size limit from sdk is 32KB => 15MB
 
+const baseEndPoint = destConfig => {
```
Comment (akashrpo) [inline v0/destinations/am/transform.js:49]:
```suggestion
const baseEndpoint = destConfig => {
```

**Review by saikumarrs — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**